lib/realtime/tenants/connect.ex

```diff
@@ -0,0 +1,121 @@
+defmodule Realtime.Tenants.Connect do
+  use GenServer
+
+  require Logger
+
+  alias Realtime.Helpers
+  alias Realtime.PostgresCdc
+  alias Realtime.Tenants
+
+  @cdc "postgres_cdc_rls"
+
+  @spec connection_status(binary()) :: {:ok, DBConnection.t()} | {:error, term()}
+  def connection_status(tenant_id) do
+    case get_status(tenant_id) do
+      :undefined -> call_external_node(tenant_id)
+      {:ok, conn} -> {:ok, conn}
+      _ -> {:error, :tenant_database_unavailable}
+    end
+  end
+
+  def set_status(tenant_id) do
+    :ok = GenServer.cast(__MODULE__, {:set_status, tenant_id})
+    set_status_backoff(tenant_id)
+  end
+
+  def get_status(tenant_id) do
+    case :syn.lookup(__MODULE__, tenant_id) do
+      {_, %{conn: conn}} when not is_nil(conn) -> {:ok, conn}
+      error -> error
+    end
+  end
+
+  def start_link(_opts), do: GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
+
+  def init(state), do: {:ok, state, {:continue, :setup_syn}}
+
+  ## GenServer callbacks
+  def handle_continue(:setup_syn, state) do
+    :ok = :syn.add_node_to_scopes([__MODULE__])
+    {:noreply, state}
+  end
+
+  def handle_cast({:set_status, tenant_id}, state) do
+    res = check_tenant_connection(tenant_id)
+    :ok = update_syn_with_conn_check(res, tenant_id)
+    {:noreply, state}
+  end
+
+  defp call_external_node(tenant_id) do
+    with tenant <- Tenants.Cache.get_tenant_by_external_id(tenant_id),
+         {:ok, node} <- Realtime.Nodes.get_node_for_tenant(tenant),
+         :ok <- :rpc.call(node, __MODULE__, :set_status, [tenant_id]) do
+      get_status(tenant_id)
+    end
+  end
```

Comment:
if a tenant's region is missing in `Realtime.Nodes.region_mapping/2`, different nodes may keep the status for the same tenant